ldap/servers/slapd/daemon.c

```diff
@@ -1901,21 +1896,27 @@ slapd_bind_local_user(Connection *conn)
                         );
 
                     if (0 == ret) {
-                        char *auth_dn = slapi_ch_strdup(
-                            slapi_entry_get_ndn(
-                                entries[0]));
-
-                        auth_dn = slapi_dn_normalize(
-                            auth_dn);
+                        char *auth_dn = NULL;
+                        char *authDNAtttr = config_get_ldapi_map_dn_attr();
+
+                        if (authDNAtttr) {
+                            /*
+                             * We configured a dn mapping attribute, now check
+                             * for it in the entry
+                             */
+                            auth_dn = slapi_entry_attr_get_charptr(entries[0], authDNAtttr);
```

Comment:
Doesn't this rely on our ldapi search base being all backends including the config be's?

Reply:
No, it always/only uses the configured LDAPI search base.  As it has always been, you have to set a single search base if you want to map entries.  It never checked "all" backends, you have to set a single backend.  So the code uses the search base to find entries based on the LDAP configuration (nsslapd-ldapiEntrySearchBase, nsslapd-ldapiUIDNumberType, etc).  Once we find entries we can change the DN to whatever backend we want (if nsslapd-ldapiauthdnattr is set), and everything works correctly.

Now your concern about dynamic users is valid, but we don't support them today.  Your proposal, which I like very much, would require a complete redesign of LDAPI entry mapping.  LDAPI, as of today, is based on static/hardcoded uidNumber/gidNumbers in entries.  This PR is still compliant with that approach of mapping uid/gid to some entry, but to do what you are suggesting would require rewriting LDAPI mapping code.  

IMHO this PR is very lightweight, it's back-portable, and does not change the behavior of how LDAPI works or is configured.  So what I think we need is an "LDAPI V2" that has its own/extended configuration, because we can't break/change how LDAPI V1 works today for customers who are actually using it and expect it to behave a certain way.

So...  I think this PR is still valid and is a good solution for the use case (which is needed now).  Then in 2.1.0 we can add dynamic user support (LDAPI-V2), and deprecate LDAPI-V1.  This allows customers to migrate to the newer design when they are ready.  

Thoughts?